### PR TITLE
feat: store `source` in `BuildError`

### PR DESCRIPTION
**Pull request by hyf0**
### Test Plan

---

### CODE REVIEW

**Comment by hyf0** [comment]
Current dependencies on/for this PR:
* main
  * **PR #315** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/315?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #316** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/316?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #317** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/317?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #319** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/319?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #320** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/320?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #321** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/321?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #322** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/322?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #323** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/323?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #324** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/324?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #325** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/325?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                      * **PR #326** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/326?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                        * **PR #327** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/327?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                          * **PR #328** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/328?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
                            * **PR #329** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/329?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                              * **PR #332** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/332?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/328?utm_source=stack-comment).

File: crates/rolldown_resolver/src/resolver.rs [hunk 1]
```diff
@@ -78,10 +78,16 @@ impl<F: FileSystem + Default> Resolver<F> {
...
-        |importer| Err(BuildError::unresolved_import(specifier.to_string(), importer.as_path())),
-      ),
+      Err(err) => {
+        if let Some(importer) = importer {
+          Err(
+            BuildError::unresolved_import(specifier.to_string(), importer.as_path())
+              .with_source(err),
+          )
+        } else {
+          Err(BuildError::unresolved_entry(specifier).with_source(err))
```
Comment (underfin) [inline crates/rolldown_resolver/src/resolver.rs:88]: Here used `with_source`, it is wried to give it `err`, maybe `with_error` is better?

Comment (hyf0) [reply crates/rolldown_resolver/src/resolver.rs:88]: `source` has it's specific meaning in error context, which is the lower-level source of this error, if any. You could refer to https://doc.rust-lang.org/std/error/trait.Error.html#method.source.

I also think about `with_cause` due to https://doc.rust-lang.org/std/error/trait.Error.html#method.cause. But it's deprected.

Comment (underfin) [reply crates/rolldown_resolver/src/resolver.rs:88]: ok

**Review by underfin — APPROVED** [review]

**Comment by hyf0** [comment]
### Merge activity

* **Nov 19, 7:13 AM**: @hyf0 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/328).
* **Nov 19, 7:18 AM**: [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/328) rebased this pull request as part of a merge.
* **Nov 19, 7:19 AM**: @hyf0 merged this pull request with [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/328).

**Pull request merged**